vsgenstubs4/init.py

```diff
@@ -0,0 +1,579 @@
+            signature = signature.replace('Union[Func, Callable]', callback_type)
+            signature = signature.replace('Union[Func, Callable, None]', f'Optional[{callback_type}]')
+
+            if plugin.namespace in special_callbacks_types:
+                for func_ns, call_type in special_callbacks_types[plugin.namespace].items():
+                    func_name, arg_name = func_ns.split('.')
+
+                    if func_name == func.name:
+                        signature = signature.replace(
+                            f'{arg_name}: {callback_type}', f'{arg_name}: VSMapValueCallback[{call_type}]'
+                        )
+                        signature = signature.replace(
+                            f'{arg_name}: Optional[{callback_type}]',
+                            f'{arg_name}: Optional[VSMapValueCallback[{call_type}]]')
+
+            # Replace the keywords with valid values
+            for kw in reserved_keywords:
+                signature = signature.replace(f' {kw}:', f' {kw}_:')
+
+            # Remove anonymous Anys
+            signature = signature.replace(' **kwargs: Any', f' **kwargs: {vs_value_type}')
+
+            # Add a self.
+            signature = signature.replace('(', '(self, ').replace(', )', ')')
+
+            yield f'    def {func.name}{signature}: ...'
+
+
+class BoundSignature:
+    def __init__(self, namespace: str, cores: Iterable[CoreLike]) -> None:
+        self.namespace = namespace
+        self.cores = cores
+
+    def __iter__(self) -> Iterator[tuple[str, Iterator[str]]]:
+        for core in self.cores:
+            signatures = retrieve_func_sigs(core, self.namespace)
+
+            try:
+                signature = next(signatures)
+            except StopIteration:
+                continue
+
+            yield core.__class__.__name__, chain.from_iterable(([signature], signatures))
+
+
+class PluginMeta(NamedTuple):
+    name: str
+    description: str
+    bound: BoundSignature
+
+    @classmethod
+    def from_namespace(cls, namespace: str, cores: Sequence[CoreLike]) -> 'PluginMeta':
+        try:
+            plugin = cores[0].__getattr__(namespace)
+        except BaseException:
+            raise ValueError(f'Invalid namespace! Plugin not found: "{namespace}"')
+
+        return PluginMeta(
+            plugin.namespace, plugin.name, BoundSignature(plugin.namespace, cores)
+        )
+
+    def _str_(self, kind: str, __x: tuple[object, ...]) -> bool:
+        return getattr(str, kind)(self.name.lower(), str(__x[0]).lower())
+
+    def __gt__(self, x: 'PluginMeta', /) -> bool: return self._str_('__gt__', x)  # type: ignore[override]
+    def __lt__(self, x: 'PluginMeta', /) -> bool: return self._str_('__lt__', x)  # type: ignore[override]
+    def __ge__(self, x: 'PluginMeta', /) -> bool: return self._str_('__ge__', x)  # type: ignore[override]
+    def __le__(self, x: 'PluginMeta', /) -> bool: return self._str_('__le__', x)  # type: ignore[override]
```

Comment:
[Why not use functools.total_odering?](https://docs.python.org/3/library/functools.html#functools.total_ordering)

Reply:
https://github.com/vapoursynth/vsrepo/pull/186#discussion_r973697159